### PR TITLE
KETTLE-59: Implementation and test cases for "resolvers", pulling environmental information into configs

### PR DESCRIPTION
**Pull request by amb26**

### CODE REVIEW

**Comment by gtirloni** [comment]
I tested this by modifying our favorite config file (gpii.preferencesServer.config.production.json) like this:

```json
{
    "type": "gpii.preferencesServer.config.production",
    "options": {
        "gradeNames": "fluid.component",
        "distributeOptions": {
            "rawPreferencesDataSource": {
                "record": {
                    "type": "kettle.dataSource.URL",
                    "options": {
                        "gradeNames": "kettle.dataSource.CouchDB",
                        "url": "@expand:kettle.resolvers.file(/tmp/couchdb_url)"
                    }
                },
                "target": "{that preferencesServer rawPreferencesDataSource}"
            },
            "logging": {
                 "record": true,
                 "target": "{that kettle.server}.options.logging"
            },
            "distributePort": {
                "record": "@expand:kettle.resolvers.env(PREFS_LISTEN_PORT)",
                "target": "{that kettle.server}.options.port"
            }
        }
    },
    "mergeConfigs": "./gpii.preferencesServer.config.base.json"
}
```

Then I created a file named `/tmp/couchdb_url` with the following content: `http://spongebob:1000/preferences/%userToken`

And started the preferences servers like this: `PREFS_LISTEN_PORT=8082 NODE_ENV=gpii.config.preferencesServer.standalone.production node gpii.js`

The expected CouchDB URL was used and the server was listening on port 8082. Awesome! :+1: :smile: 

I couldn't figure out how to work with booleans though. Changing the config like this:

```json
            "logging": {
                 "record": "@expand:kettle.resolvers.env(PREFS_LOGGING)",
                 "target": "{that kettle.server}.options.logging"
            },
```

Yielded error:

```
20:54:09.634:  FATAL ERROR: Uncaught exception: Assertion failure - check console for more details: Unrecognised fluid logging level true
```

If I removed the quotes around the resolver string, I would get an error:


```
20:57:44.671:  ASSERTION FAILED: Error: Parse error on line 17:
...          "record": @expand:kettle.resol
-----------------------^
Expecting 'STRING', 'NUMBER', 'NULL', 'TRUE', 'FALSE', '{', '[', got 'undefined'
```

**Comment by amb26** [comment]
Cheers for the testing and use cases, @gtirloni . Type-awareness of this kind you want for "logging" and others will depend on the JSON schema work underway being driven by https://issues.gpii.net/browse/GPII-2409 by @the-t-in-rtf - as this example shows, these schemas meet the needs of type conversion as well as just validation. In fact you just "got lucky" with the port number use case because the internal express plumbing converts it for you automatically, but in general you will not be so lucky.

File: docs/ConfigsAndApplications.md [hunk 1]
```diff
@@ -12,6 +12,10 @@ with the syntax and meaning of component trees, it is a good idea to browse the
 Infusion [documentation site](http://docs.fluidproject.org/infusion/development/). Kettle components are currently derived from
 the base grade `fluid.component`, so you can ignore for these purposes the parts of the Infusion documentation relating to model and view components.
 
+Note that Kettle configs need not code for Kettle applications. Their structure is
```
Comment (cindyli) [inline docs/ConfigsAndApplications.md]: What does it mean by "Kettle configs need not code for Kettle applications"? Does it mean "Kettle configs are not needed for coding Kettle applications"? Excuse my ignorance in English if this is a silly question. ;)

File: docs/ConfigsAndApplications.md [hunk 2]
```diff
@@ -73,9 +77,10 @@ You can try out these samples in [examples/simpleConfig](../examples/simpleConfi
 load a common module, `simpleConfig-client.js` which tests the server by firing an HTTP request to it and logging the payload – this uses one of the HTTP client drivers taken from Kettle's
 [testing](KettleTestingFramework.md) definitions. Later on, we will see how to issue formal test fixtures against this application by using the [Kettle testing framework](KettleTestingFramework.md).
 
-## Starting a Kettle application
+## Starting an application encoded by a Kettle config
 
-Kettle applications can be started in a variety of ways, both programmatically and from the command line - as well as easily embedded into other applications, be they Infusion component trees or raw Express apps.
+An application encoded as a Kettle config can be started in a variety of ways, 
+both programmatically and from the command line - as well as easily embedded into other applications, be they Infusion component trees or raw Express apps.
```
Comment (cindyli) [inline docs/ConfigsAndApplications.md]: typo with "be they"?

File: tests/ResolverTests.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+/**
+ * Kettle Config Loader Tests
```
Comment (cindyli) [inline tests/ResolverTests.js]: Fix the test name and the copyright.

File: lib/KettleResolvers.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+"use strict";
+
+var fluid = require("infusion"),
+    fs = require("fs"),
+    kettle = fluid.registerNamespace("kettle");
+
+fluid.registerNamespace("kettle.resolvers");
+
+kettle.resolvers.env = function (name) {
```
Comment (cindyli) [inline lib/KettleResolvers.js]: It would be nice to have code comments for these functions.

File: tests/ResolverTests.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+fluid.registerNamespace("kettle.tests");
+
+kettle.tests.loadConfigTest = function (configName) {
+    return kettle.config.loadConfig({
+        configName: configName,
+        configPath: "%kettle/tests/configs"
+    });
+};
+
+jqUnit.test("Test present environment substitution", function () {
```
Comment (cindyli) [inline tests/ResolverTests.js]: Shall the case that the requested global env variable is not defined at all be tested?

Comment (cindyli) [inline tests/ResolverTests.js]: Please ignore this comment. The following test covers this case.

File: tests/ResolverTests.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    delete process.env.KETTLE_ENV_TEST;
+});
+
+jqUnit.test("Test absent environment substitution", function () {
+    var that = kettle.tests.loadConfigTest("kettle.tests.resolver.envConfig");
+    jqUnit.assertEquals("Environment value is resolved", "OPTION6",
+        that.options.option6);
+});
+
+jqUnit.test("Test present file substitution", function () {
```
Comment (cindyli) [inline tests/ResolverTests.js]: Shall we test `kettle.tests.resolver.fileConfig` for cases:
1. the file does not exist;
2. the file exists but empty;
3. the file contains multiple lines.

File: tests/ResolverTests.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    var that = kettle.tests.loadConfigTest("kettle.tests.resolver.fileConfig");
+    jqUnit.assertEquals("Environment value is resolved", "This is a secret secret",
+        that.options.option6);
+});
```
Comment (cindyli) [inline tests/ResolverTests.js]: Missing tests for `kettle.resolvers.args`?

File: docs/ConfigsAndApplications.md (hunk 1 above)
Comment (amb26) [inline docs/ConfigsAndApplications.md]: Reworded for clarity

File: docs/ConfigsAndApplications.md (hunk 2 above)
Comment (amb26) [inline docs/ConfigsAndApplications.md]: Reworded for clarity

**Comment by cindyli** [comment]
Merged at 9ff8c535d095249b03ea299dde42c503ebe1bd99

**Comment by gtirloni** [comment]
This is great! Thank you very much!
yourock

**Pull request merged**